### PR TITLE
Easily train a new fast tokenizer from a given one

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR does two different things at the same time:
- it allows to instantiate a subclass of a `PreTrainedTokenizerFast` with just the tokenizer object by making arguments like vocab or merges optional (only done for three models here but can complete if the design is accepted)
- adds a method to train a new fast tokenizer from an existing one, using the same normalizer, pre-tokenizers and post-processors.

With this done, one can do:
```
from transformers import AutoTokenizer

checkpoint = "bert-base-cased" # or any checkpoint that has a fast tokenizer.
tokenizer = AutoTokenizer.from_pretrained(checkpoint)
assert tokenizer.is_fast, "This only works for fast tokenizers."

# Should be a generator of list of texts.
training_corpus = [
    ["This is the first sentence.", "This is the second one."],
    ["This sentence (contains #) over symbols and numbers 12 3.", "But not this one."],
]
new_tokenizer = tokenizer.train_new_from_iterator(training_corpus, vocab_size=25000)
```
The new tokenizer can then be used, saved, pushed to the hub. It has the same type as `tokenizer`.

### CODE REVIEW

File: src/transformers/tokenization_utils_fast.py [hunk 1]
```diff
@@ -555,3 +562,50 @@ def _save_pretrained(
...
+            special_tokens = []
+            for added_token in added_tokens:
+                if tokenizer_json["model"]["type"] != "Unigram":
+                    special = added_token.pop("special")
+                    if not special:
+                        continue
+                    _ = added_token.pop("id")
+                special_tokens.append(AddedToken(**added_token))
+
+        trainer_class = MODEL_TO_TRAINER_MAPPING[tokenizer_json["model"]["type"]]
```
Comment (n1t0) [inline src/transformers/tokenization_utils_fast.py:597]: You can also get and modify the trainer directly with
```python
trainer = tokenizer.model.get_trainer()
trainer.vocab_size = vocab_size
trainer.special_tokens = special_tokens
```
Another advantage of doing this, is that it should work directly when we'll be able to get the trainer with the parameters that were used for the initial training.

**Review by n1t0 — APPROVED** [review]
That's a very nice addition! Looks good to me, and I love how you handled the post-processor at the end!

File: src/transformers/tokenization_utils_fast.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/tokenization_utils_fast.py:597]: Can all kwargs accepted by the init be passed to the `get_trainer` method? The idea was to let the user customize all of them.

File: src/transformers/models/albert/tokenization_albert_fast.py [hunk 2]
```diff
@@ -121,7 +121,7 @@ class AlbertTokenizerFast(PreTrainedTokenizerFast):
 
     def __init__(
         self,
-        vocab_file,
+        vocab_file=None,
```
Comment (LysandreJik) [inline src/transformers/models/albert/tokenization_albert_fast.py:124]: This is a very welcome change!

**Review by LysandreJik — APPROVED** [review]
Cool, this is a nice feature, thanks for working on it!

Looking forward to the tests!

**Review by SaulLu — COMMENTED** [review]
Thank you very much for this PR @sgugger ! Really great all these new features. 

In addition to @n1t0 's comment, I just have some questions about the treatment of special tokens when re-training the tokenizer vocabulary. If I understood correctly, the `special_tokens` argument of the new `train_new_from_iterator` is intended to be a list of string or `AddedTokens`. If this is the case, I think we could have some problems as explained in my following comments. Let me know your thoughts :slightly_smiling_face:

File: tests/test_tokenization_common.py [hunk 3]
```diff
@@ -3144,6 +3153,27 @@ def test_special_tokens_initialization(self):
...
+        # This feature only exists for fast tokenizers
+        if not self.test_rust_tokenizer:
+            return
+
+        tokenizer = self.get_rust_tokenizer()
+        new_tokenizer = tokenizer.train_new_from_iterator(SMALL_TRAINING_CORPUS, 100)
+
+        # Test we can use the new tokenizer with something not seen during training
+        inputs = new_tokenizer(["This is the first sentence.", "This sentence is different 🤗."])
+        self.assertEqual(len(inputs["input_ids"]), 2)
```
Comment (LysandreJik) [inline tests/test_tokenization_common.py:3166]: This test is nice but it isn't really thorough. How about testing after encoding & decoding a sequence, to ensure that the full approach works?

For example here, I'm trying the following with the ALBERT tokenizer:
```py
tokenizer.decode(inputs["input_ids"][0])
```
which returns the following sequence:
```
'[CLS]-")(")-.( as" was– the£[SEP]'
```

Now I may have misunderstood something, but I would expect a somewhat correctly tokenized/untokenized sequence. I am mistaken?

Comment (sgugger) [reply tests/test_tokenization_common.py:3166]: Added a decoding test, which made me catch a problem with GPT and Herbert :-)

File: tests/test_tokenization_common.py [hunk 4]
```diff
@@ -3144,6 +3153,30 @@ def test_special_tokens_initialization(self):
...
+            return
+
+        tokenizer = self.get_rust_tokenizer()
+        new_tokenizer = tokenizer.train_new_from_iterator(SMALL_TRAINING_CORPUS, 100)
+
+        # Test we can use the new tokenizer with something not seen during training
+        inputs = new_tokenizer(["This is the first sentence", "This sentence is different 🤗."])
+        self.assertEqual(len(inputs["input_ids"]), 2)
+        decoded_input = new_tokenizer.decode(inputs["input_ids"][0], skip_special_tokens=True)
+        self.assertEqual("this is the first sentence", decoded_input.lower())
```
Comment (LysandreJik) [inline tests/test_tokenization_common.py:3168]: Absolute nitpick, but here you could check for lowercasing with
```
new_tokenizer.init_kwargs['do_lower_case']
```
and check against the actual sequence

**Review by LysandreJik — APPROVED** [review]
Great, thanks a lot @sgugger. This is a fantastic addition!

Should there be an additional document in the doc to showcase this?

File: src/transformers/tokenization_utils_fast.py [hunk 5]
```diff
@@ -60,6 +62,12 @@
             from 🤗 tokenizers <../fast_tokenizers>` for more information.
 """
 
+MODEL_TO_TRAINER_MAPPING = {
+    "BPE": BpeTrainer,
+    "Unigram": UnigramTrainer,
+    "WordPiece": WordPieceTrainer,
+}
+
```
Comment (SaulLu) [inline src/transformers/tokenization_utils_fast.py:70]:
```suggestion
MODEL_TO_TRAINER_MAPPING = {
    "BPE": BpeTrainer,
    "Unigram": UnigramTrainer,
    "WordLevel": WordLevelTrainer,
    "WordPiece": WordPieceTrainer,
}

```

Comment (SaulLu) [reply src/transformers/tokenization_utils_fast.py:70]: I have the impression that none of the fast tokenizers we test use a `WordLevel` tokenizer model, but if that were the case I think we would need this extra line.

Maybe we should add a test that uses such a tokenizer? For example, in the fast tokenizer tests?

**Review by SaulLu — COMMENTED** [review]
Love the new addtions. Just wanted to share two little things before forgetting them :slightly_smiling_face: !

File: tests/test_tokenization_common.py [hunk 6]
```diff
@@ -3144,6 +3154,93 @@ def test_special_tokens_initialization(self):
...
+                new_id = new_tokenizer.get_vocab()[new_special_token]
+                self.assertEqual(getattr(new_tokenizer, f"{token}_id"), new_id)
+
+        # Test we can use the new tokenizer with something not seen during training
+        inputs = new_tokenizer(["This is the first sentence", "This sentence is different 🤗."])
+        self.assertEqual(len(inputs["input_ids"]), 2)
+        decoded_input = new_tokenizer.decode(inputs["input_ids"][0], skip_special_tokens=True)
+        expected_result = "This is the first sentence"
+        if new_tokenizer.init_kwargs.get("do_lower_case", False):
+            decoded_input = decoded_input.lower()
```
Comment (SaulLu) [inline tests/test_tokenization_common.py:3240]: A general question: I'm not sure to understand why we need to lower case the `decoded_input` if the `do_lower_case` argument is set to `True`.

Comment (sgugger) [reply tests/test_tokenization_common.py:3240]: Yes it's just the expected result. I had it wrong and forgot to remove this one :facepalm:

Comment (SaulLu) [reply tests/test_tokenization_common.py:3240]: Got it! :+1:

**Comment by sgugger** [comment]
Failure is spurious, so merging!

**Pull request merged**